metron-platform/metron-enrichment/src/main/java/org/apache/metron/enrichment/bolt/JoinBolt.java

```diff
@@ -90,11 +93,24 @@ public void prepare(Map map, TopologyContext topologyContext, OutputCollector ou
       }
     };
     cache = CacheBuilder.newBuilder().maximumSize(maxCacheSize)
-            .expireAfterWrite(maxTimeRetain, TimeUnit.MINUTES)
+            .expireAfterWrite(maxTimeRetain, TimeUnit.MINUTES).removalListener(new JoinRemoveListener())
             .build(loader);
     prepare(map, topologyContext);
   }
 
+  class JoinRemoveListener implements RemovalListener<String, Map<String, V>> {
+
+    @Override
+    public void onRemoval(RemovalNotification<String, Map<String, V>> removalNotification) {
+      if (removalNotification.getCause() == RemovalCause.SIZE) {
+        LOG.error("Join cache reached max size limit. Increase the maxCacheSize setting or add more tasks to " + this.getClass().getSimpleName() + ".");
+      }
+      if (removalNotification.getCause() == RemovalCause.EXPIRED) {
+        LOG.error("Message was in the join cache too long which may be caused by slow enrichments/threatintels.  Increase the maxTimeRetain setting.");
```

Comment:
We should also do a `Collector.reportError()` on the message, so it pops up in the Storm UI.